Feat(Mesh): Data plane on Kubernetes (#3854)
Kong/developer.konghq.com: app/mesh/data-plane-kubernetes.md

      k8s.kuma.io/service-name: my-service
      k8s.kuma.io/service-port: "1200"
      k8s.kuma.io/namespace: my-namespace
      # Labels coming from your Pod
      app: my-app
      foo: bar
  - port: 8080
    tags:
      kuma.io/protocol: http
      kuma.io/service: my-other-service_my-namespace_svc_81
      k8s.kuma.io/service-name: my-other-service
      k8s.kuma.io/service-port: "81"
      k8s.kuma.io/namespace: my-namespace
      # Labels coming from your Pod
      app: my-app
      foo: bar
```

Notice how `kuma.io/service` is built on `<serviceName>_<namespace>_svc_<port>` and `kuma.io/protocol` is the `appProtocol` field of your Service entry.

## Capabilities

The sidecar doesn't need any [capabilities](https://kubernetes.io/docs/tasks/configure-Pod-container/security-context/#set-capabilities-for-a-container) and works with `drop: ["ALL"]`. You can use [`ContainerPatch`](#custom-container-configuration) to control capabilities for the sidecar.

## Lifecycle

The lifecycle management has been significantly improved with the `SidecarContainers` feature introduced in Kubernetes v1.29.

### Kubernetes sidecar containers

> To enable the use of the Kubernetes sidecar containers feature with {{ site.mesh_product_name }}, enable the `experimental.sidecarContainers`/`KUMA_EXPERIMENTAL_SIDECAR_CONTAINERS` option.

Reviewer: ```suggestion
To use Kubernetes sidecar containers with {{ site.mesh_product_name }}, enable the `experimental.sidecarContainers`/`KUMA_EXPERIMENTAL_SIDECAR_CONTAINERS` option.
```

{:.warning}
> The `waitForDataplaneReady` setting relies on the fact that defining a `postStart` hook causes Kubernetes to run containers sequentially based on their order of occurrence in the `containers` list.
> This isn't documented and could change in the future.
> It also depends on injecting the `kuma-sidecar` container as the first container in the Pod, which isn't guaranteed since other mutating webhooks can rearrange the containers.

### Leaving the mesh

To leave the mesh in a graceful shutdown, you need to remove the traffic destination from all the clients before shutting it down.

When the {{site.mesh_product_name}} sidecar receives a SIGTERM signal, it:

1. Starts draining Envoy listeners.
1. Waits the entire drain time.
1. Terminates.

While draining, Envoy can still accept connections, however:

1. It is marked unhealthy on the Envoy Admin `/ready` endpoint.
1. It sends `connection: close` for HTTP/1.1 requests and the `GOAWAY` frame for HTTP/2. This forces clients to close their connection and reconnect to the new instance.

Whenever a user or system deletes a Pod, Kubernetes does the following:

1. It marks the Pod as terminated.

> 1. For every container concurrently, it:

Reviewer: ```suggestion
1. For every container running concurrently, it:
```
Not sure if this was missing a word or not

* Add a pre-stop hook to postpone stopping the application container:

  ```yaml
  apiVersion: apps/v1
  kind: Deployment
  metadata:
    name: redis
  spec:
    template:
      spec:
        containers:
        - name: redis
          image: "redis"
          lifecycle:
            preStop:
              exec:
                command: ["/bin/sleep", "15"]
  ```

When a Pod is deleted, its matching `Dataplane` resource is deleted as well. This is possible thanks to the
[owner reference](https://kubernetes.io/docs/concepts/overview/working-with-objects/owners-dependents/) set on the `Dataplane` resource.

## Custom container configuration

If you want to modify the default container configuration, you can use the `ContainerPatch` Kubernetes CRD. 
It allows configuration of both sidecar and init containers. 
`ContainerPatch` resources are namespace-scoped and can only be applied in a namespace where a {{site.mesh_product_name}} control plane is running.

{:.warning}
> In most cases you shouldn't need to override the sidecar and init container configurations. 

> > `ContainerPatch` is a feature which requires good understanding of both {{site.mesh_product_name}} and Kubernetes.

Reviewer: ```suggestion
In most cases you shouldn't need to override the sidecar and init container configurations. 
`ContainerPatch` is a feature which requires good understanding of both {{site.mesh_product_name}} and Kubernetes.
```
I think this would be okay as regular text instead of a note.

  namespace: app-ns
  name: app-deployment
spec:
  replicas: 1
  selector:
    matchLabels:
      app: app-deployment
  template:
    metadata:
      labels:
        app: app-deployment
      annotations:
        kuma.io/container-patches: container-patch-1,container-patch-2
    spec: [...]
```

### Default patches

You can configure `kuma-cp` to apply the list of default patches for workloads that don't specify their own patches by modifying the `containerPatches` value from the `kuma-dp` configuration:

```yaml
[...]
runtime:
  kubernetes:
    injector:
      containerPatches: [ ]
[...]
```

{:.info}
> If you specify the list of default patches (i.e. `["default-patch-1", "default-patch-2]`) but your workload is annotated with its own list of patches (for example, `["Pod-patch-1", "Pod-patch-2]`) only the latter will be applied.

To install a control plane with environment variables, you can run:

```sh
kumactl install control-plane --env-var "KUMA_RUNTIME_KUBERNETES_INJECTOR_CONTAINER_PATCHES=patch1,patch2"
```

### Error modes and validation

When applying `ContainerPatch`, {{site.mesh_product_name}} validates that the rendered container spec meets the Kubernetes specification.

> However, {{site.mesh_product_name}} doesn't validate that it's a sane configuration.

Reviewer: ```suggestion
However, {{site.mesh_product_name}} doesn't validate that it's a sound configuration.
```

If a workload refers to a `ContainerPatch` that doesn't exist, the injection will explicitly fail and log the failure.

## Direct access to services

On Kubernetes, by default:
* Data plane proxies communicate with each other by leveraging the `ClusterIP` address of the `Service` resources. 
* Any request made to another Service is automatically load-balanced client-side by the data plane proxy that originates the request (they are load balanced by the local Envoy proxy sidecar proxy).

> There are situations where you may want to bypass the client-side load balancing and directly access services by using their IP address (for example, in the case of Prometheus wanting to scrape metrics from services by their individual IP address).

Reviewer: ```suggestion
There are situations where you may want to bypass the client-side load balancing and directly access services by using their IP address (for example, in the case of Prometheus scraping metrics from services by their individual IP address).
```

When an originating Service wants to directly consume other Services by their IP address, the originating service's `Deployment` resource must include the following annotation:

```yaml
kuma.io/direct-access-services: Service1, Service2, ServiceN
```

The value is a comma separated list of {{site.mesh_product_name}} Services that will be consumed directly. For example:

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: example-app
  namespace: kuma-example
spec:
  ...
  template:
    metadata:
      ...
      annotations:
        kuma.io/direct-access-services: "backend_example_svc_1234,backend_example_svc_1235"
    spec:
      containers:
        ...
```

{:.info}

> > When using direct access with a [headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services), the destination Service will be accessible at: `{{site.mesh_product_name}}-service.pod-name.mesh`

Reviewer: ```suggestion
> When using direct access with a [headless Service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services), the destination Service will be accessible at `{{site.mesh_product_name}}-service.pod-name.mesh`.
```

> We can also use `*` to indicate direct access to every Service in the mesh:

Reviewer: ```suggestion
You can also use `*` to indicate direct access to every Service in the mesh:
```
